OrbitControls : pan damping + zoom damping

Astrak opened the pull request:
These changes provide OrbitControls with pan damping and zoom damping (for both perspective and orthographic cameras). You can check the [current orbitcontrols example](http://rawgit.com/Astrak/three.js/orbitControls-panDamping-zoomDamping/examples/misc_controls_orbit.html), which behaviour is unchanged. [A second OrbitControls example](http://rawgit.com/Astrak/three.js/orbitControls-panDamping-zoomDamping/examples/misc_controls_orbit2.html) (that may need refinements in the scene) shows the differences, when clicking `enableDamping` in the GUI.

The API has changed with `dampingFactor` now thus being only `rotateDampingFactor`, but this is handled internally as the rest of the old APIs, so no break when migrating.

Changes : 
- `dampingFactor` > `rotateDampingFactor`
- added pan damping and the `panDampingFactor` property (since there is `panSpeed`)
- added zoom damping and the `zoomDampingFactor` property (since there is `zoomSpeed`)
- new example orbit2.html, it uses a new pattern which mixes the 2 current patterns with orbitccontrols to make the most of the 'change' event listener without damping, and of the rAF with damping.
- OrbitControls doc updated
- internally, `dollyIn` and `dollyOut` have been merged in a single function renamed `assignZoom`, and the zoom change it was handling for the Orthographic Camera case is now handled in the `update` method, so zoom damping can work for that camera too.

Notes : 
- Main motivation : UX improvement (to my mind obviously). It also makes OrbitControls' behaviour consistent with TrackballControls, that do have damping for pan/rotate/zoom, and where `staticMoving` enables/disables damping for them all.
- In the first commits, the `zoomSpeed` was modified arbitrarily to answer some specific cases when called from a mousewheel event, this is now removed (I believe it should be discussed elsewhere to simplify the PR).

Astrak committed:
damping in pan

damping in pan

Astrak committed:
modified API

added panDampingFactor and panSpeed
modified dampingFactor to rotateDampingFactor (no breaking changes)
changed native rotateDampingFactor from 0.25 to 0.1

Astrak committed:
updated doc

Astrak committed:
updated examples

Astrak committed:
handled zoom damping

dollyIn & dollyOut > just dolly
zoom damping implemented for perspective & orthographic cameras

Astrak committed:
updated doc

Astrak committed:
updated example

Astrak committed:
example : update call fix

Astrak committed:
clean + autoRotate

mrdoob commented:
@WestLangley looks good?

WestLangley commented:
Before I address some of the implementation issues here, I'd like some feedback first, please.

@mrdoob Are these features -- and parameters -- something you want to add?

Are different damping constants needed for mouse and touch? If so, I am concerned.

@mrdoob Please have a look at how the animation loop is handled in this example. It is different from the usual pattern. How do you feel about it?

On the mac, inertia when scrolling (zooming) is built in to the behavior of the OS. Is adding inertia in software going to be an issue?

I see in this example that panning does not work when zoomed all the way in.

I prefer to continue to have an example that does not require an animation loop. Adding a second example, instead, makes sense to me.

@Mugen87 Do you mind trying to break this? :)

Astrak commented:
Hey @WestLangley , thanks for the feedback, I would like to answer three of the interrogations you have :

- The need : damping to zoom seemed especially needed to me for a while, because if one implements a render flow similar to Sketchfab or Clara.io or GoogleMaps, where the render stops at camera stop (saves resources) and performs one last huge antialiasing to improve the picture, zooming in and out creates a ton of app-related 'render-end'-like events, while the camera is supposed to be in a move, and the aliasing asks too much resources for the zoom, making it laggy. This has been a real limitation for such apps to me for a long time.

- panning less when zoomed more : I did not change that, check https://threejs.org/examples/?q=orbit#misc_controls_orbit

- render loop pattern : since the GUI allows to enable damping, the render loop has to be the one with animate. Here I also chose to replace the render calls by a boolean : in this past few years it looked a better practice to me to make everything only render through the rAF, @mrdoob also recently got us aware on twitter that browsers can fire more than one mousemove event per frame. Regarding the resize event, being an intensive operation, it can also look prefered to wait for the rAF to render. For all these reasons, even without damping, I always use to stick to that pattern with the rAF as the unique entry point to a render call. Also when I began learning threejs few years ago, having two patterns added a bit of confusion to me. Finally, this pattern making also use of that boolean with the damping through the 'change' event, allows not to render the scene when nothing moves, as implemented on Sketchfab, GoogleMaps, Clara.io... In the end this is the pattern I use everywhere, whatever the parameters of the scene and I would argue to make it the unique pattern, for people to have more optimized scenes and keep everything rendered through the rAF. This pattern stills relies on the 'change' event.

Astrak committed:
reset misc_controls_orbit

reset misc_controls_orbit.html, modified dampingFactor > rotateDampingFactor
added 'no animate loop' at the top
removed unused stats

Astrak committed:
created misc_controls_orbit2

Astrak committed:
more parameters in the GUI

added polarAngle, azimuthAngle, distance - related parameters in the GUI

Astrak committed:
azimuth : max is 2*pi

Mugen87 commented:
Phew, i'm not very familiar with the code of `OrbitControls` but i generally prefer a simple code base. I like it when the classes of `three.js` provide proper implemented functionalities for the most common requirements. I think `OrbitControls` in its current state is appropriate and does not necessarily require enhancements.

@Astrak Please do not understand this as a review of your work. It is more a general statement about my personal philosophy of `three.js`s software components.

Astrak commented:
@Mugen87 , 

> i generally prefer a simple code base

Regarding the js file, it should actually be simplified. Also, this new behaviour is the one that TrackballControls use : `staticMoving` disables/enables damping for rotate/zoom/pan. (Please don't conclude to disable zoom and pan damping in OrbitControls... :sweat_smile: ). 

> OrbitControls in its current state is appropriate and does not necessarily require enhancements.
- UX improvements
- non-permanent renders problems solved

Are these unnecessary enhancements ? I've been needing this in every job I had with ThreeJS, the number of lines are the same, so you may have something in mind that I don't get here.

WestLangley commented on examples/js/controls/OrbitControls.js:
```diff
@@ -279,7 +302,7 @@ THREE.OrbitControls = function ( object, domElement ) {
 	function getZoomScale() {
 
-		return Math.pow( 0.95, scope.zoomSpeed );
+		return 0.05 * scope.zoomSpeed;
```
Let's get rid of this method and inline it.

WestLangley commented on examples/js/controls/OrbitControls.js:
```diff
@@ -302,7 +325,7 @@ THREE.OrbitControls = function ( object, domElement ) {
 
 			v.setFromMatrixColumn( objectMatrix, 0 ); // get X column of objectMatrix
-			v.multiplyScalar( - distance );
+			v.multiplyScalar( - distance * scope.panSpeed );
```
What do you think the units of panSpeed are?

Astrak replied:
Thanks for the review. Good question... `panSpeed` is a coefficient but one could formulate it as the number of GL units added to a displacement of 1 GL unit (GLunit/GLunit..) in the time elapsed since the previous `update` call ?

WestLangley replied:
That would be unit-less -- a "pan speed factor". But I wouldn't worry about the nomenclature at this point.

WestLangley commented on examples/js/controls/OrbitControls.js:
```diff
@@ -366,38 +389,15 @@ THREE.OrbitControls = function ( object, domElement ) {
 	}();
 
-	function dollyIn( dollyScale ) {
+	function dolly( dollyScale, zoomOut ) {
```
I'm not liking the zoomOut flag. Can you rewrite this using a signed quantity?

Also, maybe the first arg is a distance, not a scale?

WestLangley replied:
I was suggesting refactoring the code to remove the second argument -- not renaming it.

Shouldn't you be dollying a signed distance?

Astrak replied:
Right, that's better with just one argument. I don't know about that argument being a scale or a distance. It definitely is a distance with `PerspectiveCamera` so I renamed the variable `scaleDelta` to `distanceDelta`, but with `OrthographicCamera` it just changes the zoom. So isn't the word 'scale' more appropriate (and actually 'dolly' misleading) in an isometric view ? Anyway I renamed it `dollyDistance` too for now.

WestLangley replied:
Yes, this is sort of a mess... Create the most elegant code you can, with a live demo, and we can have @mrdoob give it a look.

Astrak replied:
`dolly ( dollyDistance )` is now `assignZoom ( zoomValue )`.

WestLangley commented on examples/js/controls/OrbitControls.js:
```diff
@@ -465,15 +465,10 @@ THREE.OrbitControls = function ( object, domElement ) {
+		//Arbitrary lower the zoom scale when damping enabled and zoom called from touch and move :
+		//the zoom happens per event, but while scrolling we call 1 event per gesture,
+		//much more happen for mousemove/touchmove. So they need to scale less.
+		dolly( getZoomScale() * ( scope.enableDamping ? 0.4 : 1 ), dollyDelta.y < 0 );
```
This worries me... The fact that you added this line means you were not happy without it.

Is this fine-tuned to the particular platform/OS you tested it against? Or does it work universally?

Astrak replied:
OrbitControls trigger one zoom change for each scroll/mousemove/touchmove event. But this is not expected, because with one finger gesture of ~1-3cm, with a same effort from the user, a mousemove/touchmove will trigger a dozen of events, resulting in a big scale change, while a mousewheel only triggers 4-5 events, resulting in a small scale change. At least on windows/linux for all I know, so setting `zoomSpeed` has always been tricky to me. Demos with orbitControls always ask a lot of mousewheels, or if they don't, then the touchmove/mousemove is too sensible.

I conclude that mousewheel should be given more value (or mousemove/touchmove less value), with an arbitrary coefficient like here, it should be the same for all OS because the problem only is about mousewheel generating fewer events. But I'm wrong to only do it with damping here, this is the same in both modes. 

You say that an action on the mousewheel on MacOS triggers much more events ? Then solving that would indeed get worrysome... But I hope it's clarified at least.

WestLangley replied:
A mouse wheel event on macOS can generate perhaps 30 events due to built-in inertia. At least that's my experience.

Astrak replied:
In last commits I propose to remove my arbitrary coefficients for mousewheel/touchmove/mousemove because it seems to deserve a specific issue/PR with more in-depth analysis.

Astrak committed:
removed getZoomScale

Astrak committed:
'zoomOut' flag replaced by 'zoomSign'

Astrak committed:
zoom scale change, based on Math.pow( 0.95, zoomSpeed ) as before

Astrak committed:
fixed zoom bug

forgot to handle the case where deltaY is 0, as handled before

Astrak committed:
touchmove/mousemove vs mousewheel

Removed the enableDamping dependency.
Replaced the two comments with the two 0.4 coeff in touchmove/mousemove with one comment and one 1.5 coeff in mousewheel.

Astrak committed:
dolly 1 arg + naming

scaleDelta renamed to distanceDelta
dollyScale renamed to dollyDistance

Astrak committed:
panSpeed location

panSpeed in handleMouseMovePan/handleTouchMovePan

Astrak committed:
removed arbitrary weight for mousewheel

this could be a more expanded discussion with related tests in another issue/PR, better remove it from this one.

Astrak committed:
semantics : dolly() to assignZoom() and arg 'dollyDistance' to 'zoomValue'

Astrak committed:
semantics : removed every 'dolly' to 'zoom', consistent with pan/rotate

zoomDelta > orthographicZoomDelta (the real 'zoom' property is only used for orthographic camera)
dollyStart > zoomStart
dollyEnd > zoomEnd
dollyDelta > zoomDelta

WestLangley commented:
I appreciate your efforts on this, but it seems to me you have added confusion with your semantics changes. I know that is an opinion.

Also, so you are aware, I don't know if this will be merged. I don't make that decision.

As I alluded to before, on my mac, gentle zooming causes the entire scene to vanish in your new example.

Astrak commented:
Is the conclusion that pan/zoom damping is not interesting, or that the PR looks too wrong ? In this second case I'm ready to change anything you want (i agree with you on the example, I'm just not adding anymore commits if it's useless !).

WestLangley commented:
I think pan/zoom damping is a reasonable feature, I just don't know if its implementation will be problematic across different devices.

@mrdoob Can you please have a look at the [new example](http://rawgit.com/Astrak/three.js/orbitControls-panDamping-zoomDamping/examples/misc_controls_orbit2.html) and provide your opinion on the added features?

Opinions from others would be appreciated, too.

Astrak commented:
If it can help, here is a breakdown that could help discuss the proposal : 
- PR about the `dollyIn` & `dollyOut` merging and renaming, but that keeps OrbitControls identical.
- PR for pan and zoom damping. This needs very few code changes. But we could benefit a separate PR to discuss the API.
- PR with a good example, depending on the chosen API.

drcmda commented:
@Astrak i know this is years too late but would you want to re-locate the pr towards https://github.com/pmndrs/three-stdlib ?
